modules/metastorage-client/src/main/java/org/apache/ignite/internal/metastorage/client/CursorImpl.java

```diff
@@ -50,19 +52,41 @@
 
     private final Iterator<T> it;
 
-    private final Function<Object, T> fn;
+    /**
+     * Constructor.
+     *
+     * @param metaStorageRaftGrpSvc Meta storage raft group service.
+     * @param initOp                Future that runs meta storage service operation that provides cursor.
+     * @param fn                    Function transforming the element of type {@link M} to the type of {@link T}.
+     */
+    public <M> CursorImpl(
+            RaftGroupService metaStorageRaftGrpSvc,
+            CompletableFuture<IgniteUuid> initOp,
+            Function<M, T> fn
+    ) {
+        this.metaStorageRaftGrpSvc = metaStorageRaftGrpSvc;
+        this.initOp = initOp;
+        this.it = new InnerIterator(e -> singleton(e), fn);
+    }
 
     /**
      * Constructor.
      *
      * @param metaStorageRaftGrpSvc Meta storage raft group service.
      * @param initOp                Future that runs meta storage service operation that provides cursor.
+     * @param internalCacheFn       Function transforming the result of {@link CursorNextCommand} to the
+     *                              {@link Iterable} of elements of type {@link M}.
+     * @param fn                    Function transforming the element of type {@link M} to the type of {@link T}.
      */
-    CursorImpl(RaftGroupService metaStorageRaftGrpSvc, CompletableFuture<IgniteUuid> initOp, Function<Object, T> fn) {
+    public <M> CursorImpl(
+            RaftGroupService metaStorageRaftGrpSvc,
+            CompletableFuture<IgniteUuid> initOp,
+            Function<Object, Iterable<M>> internalCacheFn,
```

Comment:
I don't think that we need two explicit transformation functions. Please consider approach that is used within
`new CursorImpl<>(metaStorageRaftGrpSvc, watchRes, MetaStorageServiceImpl::watchResponse),`.